### PR TITLE
[FLINK-5612] Fix GlobPathFilter not-serializable exception

### PR DESCRIPTION
**Pull request by mushketyk**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

Fixed GlobPathFilter serialization exception. As suggested in the JIRA I've made instantiation of PathMatcher's objects lazy to avoid their serialization.

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java [hunk 1]
```diff
@@ -52,8 +53,13 @@
...
 	private static final long serialVersionUID = 1L;
 
-	private final ArrayList<PathMatcher> includeMatchers;
-	private final ArrayList<PathMatcher> excludeMatchers;
+	private final List<String> includePatterns;
+	private final List<String> excludePatterns;
+
+	// Path matchers are not serializable so we are delaying their
+	// creation until they are used
+	private ArrayList<PathMatcher> includeMatchers;
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java]: these can be transient

File: flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java [hunk 2]
```diff
@@ -106,12 +112,27 @@ public boolean filterPath(Path filePath) {
 		return true;
 	}
 
+	public ArrayList<PathMatcher> getIncludeMatchers() {
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java]: these should be private

File: flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java [hunk 3]
```diff
@@ -69,8 +75,8 @@ public GlobFilePathFilter() {
 	 * @param excludePatterns glob patterns for files to exclude
 	 */
 	public GlobFilePathFilter(List<String> includePatterns, List<String> excludePatterns) {
-		includeMatchers = buildPatterns(includePatterns);
-		excludeMatchers = buildPatterns(excludePatterns);
+		this.includePatterns = Preconditions.checkNotNull(includePatterns);
+		this.excludePatterns = Preconditions.checkNotNull(excludePatterns)	;
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java]: spaces before `;` should be removed

File: flink-core/src/test/java/org/apache/flink/api/common/io/GlobFilePathFilterTest.java [hunk 4]
```diff
@@ -138,4 +141,30 @@ public void doubleStarPattern() {
...
+
+	@Test(expected = NullPointerException.class)
+	public void testExcludePatternIsNull() {
+		new GlobFilePathFilter(
+			Collections.singletonList("**"),
+			null);
+	}
+
+	@Test
+	public void testGlobFilterSerializable() throws IOException {
```
Comment (StephanEwen) [inline flink-core/src/test/java/org/apache/flink/api/common/io/GlobFilePathFilterTest.java]: How about using `CommonTestUtils.createCopySerializable()` here? That you can validate also that the serialized copy behaves valid.

File: flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java (hunk 1 above)
Comment (StephanEwen) [inline flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java]: +1 to make them transient - this documents and enforces that

**Comment by mushketyk** [comment]
Hi @StephanEwen, @zentol 

Thank you for your comments. I've update the PR accordingly.

File: flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java [hunk 5]
```diff
@@ -52,8 +53,13 @@
...
 	private static final long serialVersionUID = 1L;
 
-	private final ArrayList<PathMatcher> includeMatchers;
-	private final ArrayList<PathMatcher> excludeMatchers;
+	private final List<String> includePatterns;
+	private final List<String> excludePatterns;
+
+	// Path matchers are not serializable so we are delaying their
+	// creation until they are used
+	transient private ArrayList<PathMatcher> includeMatchers;
```
Comment (zentol) [inline flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java]: Now this is nitpicky, but i think the convention is that transient comes after private.

**Comment by zentol** [comment]
+1 to merge.

File: flink-core/src/test/java/org/apache/flink/api/common/io/GlobFilePathFilterTest.java [hunk 6]
```diff
@@ -138,4 +140,27 @@ public void doubleStarPattern() {
...
+		GlobFilePathFilter matcher = new GlobFilePathFilter(
+			Collections.singletonList("**"),
+			Collections.<String>emptyList());
+
+		CommonTestUtils.createCopySerializable(matcher);
```
Comment (StephanEwen) [inline flink-core/src/test/java/org/apache/flink/api/common/io/GlobFilePathFilterTest.java]: A doof addition here would be to check that after the serialization, the filter behaves in the same way as before.

File: flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java (hunk 5 above)
Comment (StephanEwen) [inline flink-core/src/main/java/org/apache/flink/api/common/io/GlobFilePathFilter.java]: Agreed, let's get the style aligned with the rest of the project.

**Comment by mushketyk** [comment]
Hi @StephanEwen, @zentol 

Thank you for helping with this PR.
I've updated it according to your review.

**Comment by EronWright** [comment]
Just hit this issue, was about to file an issue.   Awesome to see this fixed.    

It is odd to me that `GlobFilePathFilter` is marked `@Internal`.

**Comment by StephanEwen** [comment]
@EronWright Agreed, this should be `@Public(Evolving)`.

**Comment by StephanEwen** [comment]
Looks good, will merge this and fix the public annotation on the fly..

**Comment by mushketyk** [comment]
Thank you @StephanEwen !

**Comment by zentol** [comment]
@mushketyk The changes have been merged, could you close this PR?

**Comment by mushketyk** [comment]
@zentol Sure. Thank you for merging this!

**Pull request closed without merging**